Mathlib/RingTheory/HahnSeries.lean

```diff
@@ -618,22 +622,109 @@ theorem order_one [MulZeroOneClass R] : order (1 : HahnSeries Γ R) = 0 := by
   · exact order_single one_ne_zero
 #align hahn_series.order_one HahnSeries.order_one
 
-instance [NonUnitalNonAssocSemiring R] : Mul (HahnSeries Γ R) where
-  mul x y :=
-    { coeff := fun a =>
-        ∑ ij in addAntidiagonal x.isPWO_support y.isPWO_support a, x.coeff ij.fst * y.coeff ij.snd
-      isPWO_support' :=
+end HahnSeries
+
+/-- We introduce a type alias for HahnSeries in order to work with scalar multiplication by series.
+If we wrote a `SMul (HahnSeries Γ R) (HahnSeries Γ V)` instance, then when `V = HahnSeries Γ R`, we
+would have two different actions of `HahnSeries Γ R` on `HahnSeries Γ V`.  See
+`Mathlib.Data.Polynomial.Module` for more discussion on this problem. -/
+@[nolint unusedArguments]
+def HahnModule (Γ R V : Type*) [PartialOrder Γ] [Zero V] [SMul R V] :=
+  HahnSeries Γ V
+
+namespace HahnModule
+
+@[ext]
+theorem ext {Γ R V} [PartialOrder Γ] [Zero V] [SMul R V]
+    (x y : HahnModule Γ R V) (coeff : x.coeff = y.coeff) : x = y := by
+  simp_all only [HahnSeries.coeff_inj]
+
+variable {V : Type*} [AddCommMonoid V]
```

Comment:
I think every declaration in this namespace assumes `[SMul R V]`. Some other variables also occur a lot. Can you please see if you can move some of these "background assumptions" to this `variable` line?

Reply:
It looks like `[SMul R V]` doesn't like to be mixed with `[SMulWithZero R V]` or `[SMulZeroClass R V]`, so I changed the `V` to `W` in the later results.